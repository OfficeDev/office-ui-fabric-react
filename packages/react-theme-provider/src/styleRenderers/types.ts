export type RuleSet = {};
export type KeyFrames = {};
export type FontFace = {};

export type StyleRendererOptions = {
  rtl: boolean;
  targetWindow: Window;
};

export interface StyleRenderer {
  renderStyles: <TRuleSet extends RuleSet>(
    ruleSet: TRuleSet,
    options: StyleRendererOptions,
  ) => { [key in keyof TRuleSet]: string };

  renderKeyframes: (keyframes: KeyFrames, options: StyleRendererOptions) => void;

  renderFontFace: (fontFace: FontFace, options: StyleRendererOptions) => void;
}
