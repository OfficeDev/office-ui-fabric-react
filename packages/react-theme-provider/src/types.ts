/**
 * A set of states for each color plate to use.
 */
export type ThemeStateSet =
  | Partial<{
      default: string;
      hover: string;
      active: string;
      disabled: string;
    }>
  | string;

/**
 * A baseline set of color plates.
 */
export type ThemePlateSet = Partial<{
  fill: ThemeStateSet;
  text: ThemeStateSet;
  subText: ThemeStateSet;
  link: ThemeStateSet;
  divider: ThemeStateSet;
  [key: string]: ThemeStateSet;
}>;

export type TokenSetType = string | { [key: string]: TokenSetType | undefined };

/**
 * Recursive partial type.
 */
export type RecursivePartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? RecursivePartial<U>[]
    : T[P] extends object
    ? RecursivePartial<T[P]>
    : T[P];
};

/**
 * A basic prepared (fully qualitied) theme object.
 */
export interface ThemePrepared {
  tokens: {
    body: ThemePlateSet;
    [key: string]: TokenSetType;
  };

  stylesheets: string[];
}

/**
 * A partial theme, provided by the customer. The internal `createTheme` helper will fill in the rest.
 */
export interface Theme extends RecursivePartial<ThemePrepared> {}
