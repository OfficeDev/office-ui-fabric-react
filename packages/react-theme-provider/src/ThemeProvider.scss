/**
 * The ThemeProvider has a root consuming class which will apply the current
 * token values to the container DIV, so that fonts and default colors, which
 * refer to token values, can be applied.
 */
.root {
  background: var(--site-body-fill);
  color: var(--site-body-text);
  font-family: var(--site-body-fontFamily);
  font-weight: var(--site-body-fontWeight);
  font-size: var(--site-body-fontSize);
  line-height: var(--site-body-fontLineHeight);

  > * {
    box-sizing: border-box;
  }
}
