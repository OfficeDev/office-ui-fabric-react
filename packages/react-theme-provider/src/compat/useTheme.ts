import { useContext } from 'react';
import { useCustomizationSettings } from '@fluentui/utilities';
import { createTheme } from '@fluentui/theme';
import { ThemeContext } from './ThemeContext';
import { Theme } from './types';

/**
 * Get theme from CustomizerContext or Customizations singleton.
 */
function useCompatTheme(): Theme | undefined {
  return useCustomizationSettings(['theme']).theme;
}

/**
 * React hook for programmatically accessing the theme.
 */
export const useTheme = (): Theme => {
  const theme = useContext(ThemeContext);
  const legacyTheme = useCompatTheme();

  return theme || legacyTheme || createTheme({});
};
