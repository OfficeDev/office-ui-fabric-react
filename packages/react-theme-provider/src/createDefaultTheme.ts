import { Theme, Tokens } from './types';
import { createTheme } from '@uifabric/styling';
import { getTokens } from './getTokens';

/**
 * Creates default theme (Fluent theme).
 */
export const createDefaultTheme = (): Theme => {
  const defaultTheme: Theme = createTheme({});
  defaultTheme.tokens = getTokens(defaultTheme);

  return defaultTheme;
};

export const defaultTokens: Tokens = {
  body: {},

  accent: {
    background: '#0078d4',
    contentColor: '#ffffff',
    borderColor: 'transparent',
    iconColor: 'inherit',

    disabled: {
      background: '#f3f2f1',
      contentColor: '#c8c6c4',
      borderColor: 'var(--accent-disabled-background)',
      iconColor: 'var(--accent-disabled-contentColor)',
    },
    hovered: {
      background: '#106ebe',
      contentColor: 'var(--accent-contentColor)',
      borderColor: 'var(--accent-borderColor)',
      iconColor: 'var(--accent-iconColor)',
    },
    pressed: {
      background: '#005a9e',
      contentColor: 'var(--accent-contentColor)',
      borderColor: 'var(--accent-borderColor)',
      iconColor: 'var(--accent-iconColor)',
    },
    focused: {
      background: 'var(--accent-background)',
      borderColor: 'var(--accent-borderColor)',
      contentColor: 'var(--accent-contentColor)',
      iconColor: 'var(--accent-iconColor)',
    },
  },

  button: {
    size: {
      smallest: '8px',
      smaller: '16px',
      small: '24px',
      regular: '32px',
      large: '40px',
      larger: '48px',
      largest: '64px',
    },
    padding: '0 16px',
    height: 'var(--button-size-regular)',
    minHeight: 'var(--button-size-regular)',
    contentGap: '10px',
    iconSize: '16px',
    borderRadius: '2px',
    borderWidth: '1px',

    fontFamily: `'Segoe UI', 'Helvetica Neue', 'Apple Color Emoji', 'Segoe UI Emoji', Helvetica, Arial, sans-serif`,
    fontSize: '14px',
    fontWeight: 'normal',

    focusColor: '#000',
    focusInnerColor: '#fff',

    background: 'white',
    borderColor: '#8a8886',
    contentColor: '#2c2621',
    iconColor: 'inherit',

    dividerColor: 'rgba(0, 0, 0, 0.1)',
    dividerLength: 'var(--button-minHeight)',
    dividerThickness: 'var(--button-borderWidth)',

    disabled: {
      background: '#f3f2f1',
      borderColor: 'var(--button-disabled-background)',
      contentColor: '#c8c6c4',
      iconColor: 'var(--button-disabled-contentColor)',
    },
    hovered: {
      background: '#f3f2f1',
      borderColor: 'var(--button-borderColor)',
      contentColor: 'var(--button-contentColor)',
      iconColor: 'var(--button-iconColor)',
    },
    pressed: {
      background: '#edebe9',
      borderColor: 'var(--button-borderColor)',
      contentColor: 'var(--button-contentColor)',
      iconColor: 'var(--button-iconColor)',
    },
    focused: {
      background: 'var(--button-background)',
      borderColor: 'var(--button-borderColor)',
      contentColor: 'var(--button-contentColor)',
      iconColor: 'var(--button-iconColor)',
    },
  },
};
