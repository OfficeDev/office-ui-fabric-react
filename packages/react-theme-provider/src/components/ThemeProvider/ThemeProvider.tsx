import * as React from 'react';
import { ThemeProviderProps } from './ThemeProvider.types';
import { useCreateTheme, useThemeProvider, themeProviderShorthandProps } from './useThemeProvider';
import { ThemeContext } from '@fluentui/react-shared-contexts';
import { useThemeProviderStyles } from './useThemeProviderStyles';
import { getSlots } from '@fluentui/react-utilities';

const InternalStyles = React.forwardRef<HTMLElement, ThemeProviderProps>((props: ThemeProviderProps, ref) => {
  const state = useThemeProvider(props, ref);
  useThemeProviderStyles(state);
  const { slots, slotProps } = getSlots(state, themeProviderShorthandProps);
  return (
    <slots.root ref={ref} {...slotProps.root}>
      {state.children}
    </slots.root>
  );
});

export const ThemeProvider = React.forwardRef<HTMLElement, ThemeProviderProps>((props, ref) => {
  const [theme] = useCreateTheme(props.theme, props.style);
  return (
    <ThemeContext.Provider value={theme}>
      <InternalStyles ref={ref} {...props} />
    </ThemeContext.Provider>
  );
});

ThemeProvider.displayName = 'ThemeProvider';
