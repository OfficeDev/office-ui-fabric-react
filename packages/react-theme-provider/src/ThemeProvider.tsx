import * as React from 'react';
import cx from 'classnames';
import { useStylesheet } from '@fluentui/react-stylesheets';
import * as classes from './ThemeProvider.scss';
import { variablesToStyleObject } from './variablesToStyleObject';
import { ThemeContext } from './ThemeContext';
import { Theme, ThemePrepared } from './Theme';
import { createTheme } from './createTheme';
import { useTheme } from './useTheme';

export type ThemeColorSet =
  | Partial<{
      default: string;
      hover: string;
      active: string;
      disabled: string;
    }>
  | string;

export type ThemePlateSet = Partial<{
  fill: ThemeColorSet;
  text: ThemeColorSet;
  subText: ThemeColorSet;
  link: ThemeColorSet;
  divider: ThemeColorSet;
}>;

export type TokenSetType = string | { [key: string]: TokenSetType | undefined };

export interface ThemeProviderProps extends React.AllHTMLAttributes<{}> {
  theme?: Theme;
}

export const ThemeProvider = React.forwardRef<HTMLDivElement>(
  (
    { theme, className, style, ...rest }: React.PropsWithChildren<ThemeProviderProps>,
    ref: React.Ref<HTMLDivElement>,
  ) => {
    const parentTheme = useTheme();
    const fullTheme = React.useMemo<ThemePrepared>(() => createTheme(parentTheme, theme), [parentTheme, theme]);
    const inlineStyle = React.useMemo<React.CSSProperties>(
      () => variablesToStyleObject(fullTheme.tokens.site, undefined, { ...style }),
      [theme, style],
    );

    // Register stylesheets as needed.
    useStylesheet(fullTheme.stylesheets);

    // Provide the theme in case it's required through context.
    return (
      <ThemeContext.Provider value={fullTheme}>
        <div {...rest} ref={ref} className={cx(className, classes.root)} style={inlineStyle} />
      </ThemeContext.Provider>
    );
  },
);
