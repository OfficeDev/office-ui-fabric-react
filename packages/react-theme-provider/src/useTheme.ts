import { useContext } from 'react';
import { useCustomizationSettings } from '@uifabric/utilities';
import { ITheme } from '@uifabric/styling';
import { ThemeContext } from './ThemeContext';
import { Theme } from './types';
import { createDefaultTheme } from './createDefaultTheme';

/**
 * React hook for programatically accessing the theme.
 */
export const useTheme = (): Theme => {
  const theme = useContext(ThemeContext);
  const legacyTheme = useLegacyTheme();
  if (theme) {
    return theme;
  }

  if (legacyTheme) {
    return legacyTheme;
  }

  return createDefaultTheme();
};

/**
 * Get theme from CustomizerContext or Customizations singleton.
 */
function useLegacyTheme(): ITheme {
  return useCustomizationSettings(['theme']).theme;
}
