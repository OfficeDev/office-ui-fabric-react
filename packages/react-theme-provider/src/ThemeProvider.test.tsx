import * as React from 'react';
import { ThemeProvider } from './ThemeProvider';
import * as renderer from 'react-test-renderer';
import { Theme, PartialTheme } from './types';
import { useTheme } from './useTheme';
import { mount } from 'enzyme';
import { mergeThemes } from '@fluentui/theme';
import { createDefaultTheme } from './createDefaultTheme';

const lightTheme = mergeThemes({
  stylesheets: [],
  tokens: {
    body: {
      background: 'white',
      contentColor: 'black',
    },
  },
});

const darkTheme = mergeThemes({
  tokens: {
    body: {
      background: 'black',
      contentColor: 'white',
    },
  },
});

describe('ThemeProvider', () => {
  it('renders a div', () => {
    const component = renderer.create(<ThemeProvider>Hello</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('can handle a partial theme', () => {
    const partialTheme: PartialTheme = {
      tokens: {
        foo: {
          background: 'red',
        },
      },
    };

    const component = renderer.create(<ThemeProvider theme={partialTheme}>Hello</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a div with styling', () => {
    const component = renderer.create(<ThemeProvider theme={lightTheme}>Hello</ThemeProvider>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders nested themes', () => {
    const component = renderer.create(
      <ThemeProvider theme={lightTheme}>
        <div>Light theme</div>
        <ThemeProvider theme={darkTheme}>
          <div>Dark theme</div>
        </ThemeProvider>
      </ThemeProvider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('provide the theme through context', () => {
    let resolvedTheme: Theme | undefined = undefined;
    const TestComponent = () => {
      resolvedTheme = useTheme();
      return null;
    };

    mount(
      <ThemeProvider theme={lightTheme}>
        <TestComponent />
      </ThemeProvider>,
    );

    const expectedTheme = mergeThemes(createDefaultTheme(), lightTheme);
    expect(resolvedTheme).toEqual(expectedTheme);
  });

  it('can apply body theme to root element', () => {
    expect(document.body.className).toBe('');
    const component = renderer.create(
      <ThemeProvider className="foo" theme={darkTheme} applyTo="element">
        app
      </ThemeProvider>,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();

    expect(document.body.className).toBe('');
  });

  it('can apply body theme to body', () => {
    expect(document.body.className).toBe('');
    const testClass = 'foo';
    const TestComponent = (
      <ThemeProvider className={testClass} theme={darkTheme} applyTo="body">
        app
      </ThemeProvider>
    );

    const wrapper = mount(TestComponent);
    const rootClass = wrapper.find('div').prop('className');
    expect(rootClass).toBe(`${testClass} ${document.body.className}`);

    wrapper.unmount();

    expect(document.body.className).toBe('');

    const component = renderer.create(TestComponent);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
