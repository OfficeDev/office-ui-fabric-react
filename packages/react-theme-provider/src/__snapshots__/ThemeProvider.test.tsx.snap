// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider renders a div 1`] = `
<div
  className=""
  style={Object {}}
>
  Hello
</div>
`;

exports[`ThemeProvider renders a div with styling 1`] = `
<div
  className=""
  style={Object {}}
>
  Hello
</div>
`;

exports[`ThemeProvider renders nested themes 1`] = `
<div
  className=""
  style={Object {}}
>
  <div>
    Light theme
  </div>
  <div
    className=""
    style={Object {}}
  >
    <div>
      Dark theme
    </div>
  </div>
</div>
`;
